omegaconf/omegaconf.py

```diff
@@ -867,3 +914,40 @@ def _select_one(
         assert False
 
     return val, ret_key
+
+
+def _make_hashable(key: Tuple[Any, ...]) -> Tuple[Any, ...]:
+    """
+    Ensure `key` is hashable.
+
+    This is achieved by turning into tuples the lists and dicts that may be
+    stored within `key`.
+    Note that dicts are sorted, so that two dicts ordered differently will
+    lead to the same resulting hashable key.
+
+    :return: a hashable version of `key`.
+    """
+    # Hopefully it is already hashable and we have nothing to do!
+    try:
+        hash(key)
+    except TypeError:
+        pass
+    else:
+        return key
+
+    new_key: List[Any] = []  # will store the new key elements
+    hashable_item: Any
+    for idx, item in enumerate(key):
+        if item is None or isinstance(item, (int, float, bool, str)):
+            hashable_item = item
+        elif isinstance(item, list):
+            hashable_item = _make_hashable(tuple(item))
+        elif isinstance(item, tuple):
+            hashable_item = _make_hashable(item)
+        elif isinstance(item, dict):
+            # We sort the dictionary so that the order of keys does not matter.
+            hashable_item = _make_hashable(tuple(sorted(item.items())))
+        else:
+            raise NotImplementedError(f"type {type(item)} cannot be made hashable")
+        new_key.append(hashable_item)
```

Comment:
Can you refresh my memory with why we need this make_hashable thing?

Reply:
This is so that the resolver cache (which is a dictionary, and thus requires hashable keys) can work with list and dictionary arguments. We didn't need it before because all resolver args were strings.

Reply:
This function is complex and bugs in it will result in very subtle behavior changes that will be very hard to identify.
Do you have dedicated tests for it?

Reply:
> Do you have dedicated tests for it?

That would be https://github.com/odelalleau/omegaconf/blob/40976e847ea14e33a05418e4371994940ba9bbcf/tests/test_interpolation.py#L436

You're right though it's more complex than it should, I just pushed a simpler version in f468aa8